tangelo/toolboxes/molecular_computation/fno.py

```diff
@@ -0,0 +1,422 @@
+            self.threshold = threshold
+        else:
+            raise ValueError(f"The threshold {threshold} is invalid, it must be a float or a sequence of floats between 0 and 1.")
+
+        self.threshold = threshold
+
+        if self.uhf:
+            self.n_occupied = [len(x+y) for x, y in zip(self.sqmol.frozen_occupied, self.sqmol.active_occupied)]
+            self._compute_ump2_densities()
+        else:
+            self.n_occupied = len(self.sqmol.frozen_occupied + self.sqmol.active_occupied)
+            self._compute_rmp2_densities()
+
+        self.compute_fno(self.threshold)
+
+    @property
+    def fermionic_hamiltonian(self) -> FermionOperator:
+        """Property that returns a FNO fermionic hamiltonian object, with the
+        truncated active space and updated MO coefficients.
+
+        Returns:
+            FermionOperator: Self-explanatory.
+        """
+        frozen_orbitals = self.get_frozen_indices()
+        sqmol_updated = self.sqmol.freeze_mos(frozen_orbitals, inplace=False)
+        return sqmol_updated._get_fermionic_hamiltonian(self.mo_coeff)
+
+    def compute_fno(self, threshold: Union[float, List[float]]) -> None:
+        """Method to compute and truncate the FNO orbitals. It calls
+        the `_compute_rfno` or the `_compute_ufno`method, whichever is
+        appropriate.
+        """
+        self._compute_ufno(threshold) if self.uhf else self._compute_rfno(threshold)
+
+    def get_frozen_indices(self) -> Union[List[int], List[List[int]]]:
+        """Method to determine the indices of the frozen orbitals, and it calls
+        the `_get_restricted_frozen_indices` or the `_get_unrestricted_frozen_indices`
+        method, whether is appropriate.
+        """
+        return self._get_unrestricted_frozen_indices() if self.uhf else self._get_restricted_frozen_indices()
+
+    def _get_restricted_frozen_indices(self) -> List[int]:
+        """Method to determine the indices of the frozen orbitals in a
+        restricted calculation.
+
+        Returns:
+            List[int]: List containing the frozen orbital indices for the
+                molecular orbitals.
+        """
+
+        # All are set to False.
+        moidx = np.zeros(self.n_mos, dtype=bool)
+
+        # True starting from first index until an active orbital index.
+        active_virt_fno = self.n_occupied + self.n_active_virt_fno
+        moidx[: active_virt_fno] = True
+
+        # Set frozen occupied orbital to False.
+        moidx[self.frozen_occupied] = False
+
+        # Obtain the frozen indices.
+        frozen_indices = np.where(moidx == 0)[0].tolist()
+
+        return frozen_indices
+
+    def _get_unrestricted_frozen_indices(self) -> List[List[int]]:
+        """Method to determine the indices of the frozen orbitals in an
+        unrestricted calculation.
+
+        Returns:
+            List[List[int]]: List containing the frozen orbital indices for
```

Comment:
List[List[int]] to me means a list of unknown length containing other lists of unknown lengths. But it sounds like we should maybe return a 2-tuple of List[Int] instead because its very clear.

Making that change shouldn't be difficult since both items support indexing and can be unpacked the same way in the rest of your code.